cmake/Modules/CodeCoverage.cmake

```diff
@@ -139,6 +139,60 @@ function(ADD_CODE_COVERAGE)
+      # create hidden meta target which depends on hidden test targets which depend on clean_test_results
+      add_custom_target(_run_tests_${PROJECT_NAME})
+      # run_tests depends on this hidden target hierarchy to clear test results before running all tests
+     add_dependencies(run_tests _run_tests_${PROJECT_NAME})
+    endif()
+
+    if(NOT DEFINED CATKIN_ENABLE_TESTING OR CATKIN_ENABLE_TESTING)
+      # create python base coverage directory
+      add_custom_target(
+        create_python_base_coverage_dir "${CMAKE_COMMAND}" "-E" "make_directory" ${PROJECT_BINARY_DIR}/python_base_coverage
+      )
+
+      # create python base coverage report
+      add_custom_target(run_tests_${PROJECT_NAME}_python_base_coverage_report
+        COMMAND rosrun code_coverage generate_base_coverage.py ${PROJECT_SOURCE_DIR} --output ${PROJECT_BINARY_DIR}/python_base_coverage
+        COMMAND ${PYTHON_COVERAGE_PATH} report ${INCLUDE_FLAGS} ${OMIT_FLAGS} || echo "WARNING: No python base report to output"
+        COMMAND ${PYTHON_COVERAGE_PATH} xml  -o ${Coverage_NAME}_base_python.xml ${INCLUDE_FLAGS} ${OMIT_FLAGS} || echo "WARNING: No base python xml to output"
+        COMMAND ${CMAKE_COMMAND} -E copy ${PROJECT_BINARY_DIR}/python_base_coverage/${Coverage_NAME}_base_python.xml ${PROJECT_BINARY_DIR}/ || echo "WARNING: No base python xml to copy"
+        DEPENDS create_python_base_coverage_dir
+        WORKING_DIRECTORY ${PROJECT_BINARY_DIR}/python_base_coverage
+      )
+      # hidden test target which depends on building all tests and cleaning test results
+      add_custom_target(_run_tests_${PROJECT_NAME}_python_base_coverage_report
+        COMMAND rosrun code_coverage generate_base_coverage.py ${PROJECT_SOURCE_DIR} --output ${PROJECT_BINARY_DIR}/python_base_coverage
+        COMMAND ${PYTHON_COVERAGE_PATH} report ${INCLUDE_FLAGS} ${OMIT_FLAGS} || echo "WARNING: No python base report to output"
+        COMMAND ${PYTHON_COVERAGE_PATH} xml  -o ${Coverage_NAME}_base_python.xml ${INCLUDE_FLAGS} ${OMIT_FLAGS} || echo "WARNING: No base python xml to output"
+        COMMAND ${CMAKE_COMMAND} -E copy ${PROJECT_BINARY_DIR}/python_base_coverage/${Coverage_NAME}_base_python.xml ${PROJECT_BINARY_DIR}/ || echo "WARNING: No base python xml to copy"
+        DEPENDS create_python_base_coverage_dir
+        WORKING_DIRECTORY ${PROJECT_BINARY_DIR}/python_base_coverage
+      )
+    else()
+      add_custom_target(run_tests_${PROJECT_NAME}_python_base_coverage_report
+          COMMAND "${CMAKE_COMMAND}" "-E" "echo" "Skipping python base coverage report target." )
+      add_custom_target(_run_tests_${PROJECT_NAME}_python_base_coverage_report
+          COMMAND "${CMAKE_COMMAND}" "-E" "echo" "Skipping python base coverage report target." )
+    endif()
+    add_dependencies(run_tests_${PROJECT_NAME} run_tests_${PROJECT_NAME}_python_base_coverage_report)
+    add_dependencies(_run_tests_${PROJECT_NAME} _run_tests_${PROJECT_NAME}_python_base_coverage_report)
```

Comment:
I have a few questions about these two targets.
`_run_tests_${PROJECT_NAME}` is depended by `run_tests` and it looks fine.

1. `run_test_${PROJECT_NAME}` is not dependened by other targets. Could you tell me why we need it? What does `    # add target for non-test repo` mean?
2. Can these targets run at the same time? If so, `generate_base_coverage.py` and other script may run simultaneously. These targets share the same output files and I think it makes some problems with parallel building.

Reply:
I read `catkin_run_tests_target` function and find the targets.
https://github.com/ros/catkin/blob/ff31d451ce1c68d47bc058a4693aad6c0fb63a43/cmake/test/tests.cmake#L122-L191

>run_test_${PROJECT_NAME} is not dependened by other targets. Could you tell me why we need it? What does     # add target for non-test repo mean?

In order to run the test for the project, we need to set `run_tests_${PROJECT_NAME}` target.
I think this target will be executed from `catkin_tools`, but I'm not super sure.

If there is not test in the repository, `gitai_add_test_coverage` will fail because there is not tests.
But I register `run_tests_${PROJECT_NAME}_python_base_coverage_report` target for tests,
so catkin runs the `generate_coverage_report.py` as test.
This means that non-test repo has now run  `run_tests_${PROJECT_NAME}_python_base_coverage_report`,
and create basic coverage report as test.

> Can these targets run at the same time? If so, generate_base_coverage.py and other script may run simultaneously. These targets share the same output files and I think it makes some problems with parallel building.

`_run_tests_${PROJECT_NAME}` is to clean the test result and run tests.
https://github.com/ros/catkin/commit/f931db5c8c14475a9d74ffc65b9dbbe45c98d11d
these two target do not run simaltaneously.

Reply:
Good. Could you add your answers to my questions as comments in the source code briefly?

Reply:
I added the comment in cmake file.
[`6d6fa2a` (#6)](https://github.com/GITAI/code_coverage/pull/6/commits/6d6fa2acf40796d6d91d4f761a63a760abeeafbd)